ToolChain/Assembler/Assembler.py

```diff
@@ -56,79 +50,90 @@ def __init__(self, inputFile=None, outputFile=None):
-                fileContent += b"<refName>" + reference.encode("utf-8") + b"</refName>"
-                fileContent += b"<refAdd>" + struct.pack(">I", self.parser.referenceDict[reference]) + b"</refAdd>"
-
-                # Update global offsets would it be required
-                # This is required since all global offsets are set to 0 at the moment they are parsed
-                wouldBeExported = EXPORTED_REFERENCE_INDICATOR + " " + reference
-                if wouldBeExported in self.parser.referenceDict.keys():
-                    self.parser.referenceDict[wouldBeExported] = self.parser.referenceDict[reference]
-
-        fileContent += b"</InternalSymbols>"
-
-        # External symbols allow for a file to be linked with another file
-        fileContent += b"<ExternalSymbols>"
-        for reference in self.parser.referenceDict:
-            if EXPORTED_REFERENCE_INDICATOR in reference:
-                isValid = False
-                shortReference = reference.split()[1]
-                for internalReference in validInternalRef:
-                    if shortReference == internalReference:
-                        isValid = True
-                        break
-
-                if isValid:
-                    refName = reference.split()[1]
-                    fileContent += b"<refName>" + refName.encode("utf-8") + b"</refName>"
-                    fileContent += b"<refAdd>" + struct.pack(">I", self.parser.referenceDict[reference]) + b"</refAdd>"
+
+        parser = Parser()                    # Parser object which will return the binary code for instructions
+        masterString = b""                   # String that will eventually be our output file
+        xmlstring = b""                      # String that will contain the "xml" data at the beginning of the .o file
+        globalList = []                      # Temporary list to hold the names of global(external) symbols
+        externalList = []                    # Master list of all external symbols to be used by linker
+        internalList = []                    # Master list of all internal symbols
+        offset = 0                           # Offset used to calculate a label's offset from the beginning of the file
+        lineno = 0                           # Keeps track of which line we're parsing to pinpoint errors
+
+        operand = 0                          # Flag for an operand, meaning parser found a valid instruction
+        localReference = 1                   # Flag for local function, internal symbol
+        globalReference = 2                  # Flag for external function, global symbol
+        comment = 3                          # Flag for comment
+
+        file = open(inputFile, mode="r")
+        fileLines = file.readlines()
+        file.close()
+
+        masterString += b"<Text>"
+
+        for line in fileLines:
+            lineno += 1
+            # If the file isn't empty, process the content
+            if line[0] != "\n":
```

Comment:
There is a bug here.
If someone writes an assembly file that contain a line where there is a space followed by a new line the line is viewed as valid and containing actual code. I modified a file to create this case and I got the following error message:

```
$ python3 Assembler.py -i hello.casm -o hello.o
Assembler about to begin, following options will be used
  input file:             hello.casm
  output file:            hello.o
Traceback (most recent call last):
  File "Assembler.py", line 102, in <module>
    assembler = Assembler(usableArgs.input, usableArgs.output)
  File "/home/user/Desktop/Sparta/spartacus/ToolChain/Assembler/Assembler.py", line 53, in __init__
    self._parseAssembledFile(inputFile, outputFile)
  File "/home/user/Desktop/Sparta/spartacus/ToolChain/Assembler/Assembler.py", line 95, in _parseAssembledFile
    instruction, offset, labelFlag = parser.parse(line)
  File "/home/user/Desktop/Sparta/spartacus/ToolChain/Assembler/Parser/Parser.py", line 80, in parse
    dataIdentifier = line[0].upper()        # used for special identifiers such as .dataAlpha, comments etc.
IndexError: list index out of range

```

Reply:
The debugger was something I left out accidentally when re-changing the register names. That'll be a simple fix. I'll look into fixing the space + newline, and the immediate conversion problems as well